README.md
# Project: Scape
Project: Scape. A minimalistic take on the classic version of popular game Runescape.

Primarily intended as a learning experience, I tried to avoid using major packages and especially game related ones that would trivialise the work needed. Since I don't plan on working with games much in the future, anything I learn about those specific packages would be useless. Instead I decided to build my own basic systems from the ground up, even if they aren't the best possible implementations, just to get a better feel for what is possible and efficient with basic python. 

The largest package that is heavily used is Tkinter. While originally envisioned as a text/console based project, I quickly decided it would be more realistic to have at least a basic GUI in 2021. In addition, many core modules like time, random and threading were used out of necessity.

Again, due to the nature of the project, the quality of code throughout will vary depending on the time at which it was written. While I could go back and change a lot of the early design choices, it would take too much time for little gain compared to moving on to something new. In a similar vein the formatting of the code may not be perfectly in line with common standards, since I just relied on Pycharm warnings and recommendations throughout most of the process before discovering that this does not match the guidelines in their entirety.

The structure of the project allows me to easily add more in future if I want to test out something new or while learning more advanced methods. For example, I only added the account system late in development to play around with SQLite, and it wasn't too difficult to get things working again. 

As a final note before moving on to the actual gameplay: the world building was based heavily on the original Runescape to save time, but with some changes where I felt they were needed or interesting. The game's core mechanics (like how skills are trained) were intended to follow the source material where possible, but obviously being made from scratch due to the very different style of gameplay, and with some creative liberties at times. The main benefit of this approach was the existence of clear goals, since I am very goal-oriented and this gave me a large list of mini projects to work through from the beginning.


# User Experience / How to play
Interaction with the game is done entirely through button presses which are presented to the user depending on their current progress and activity. Feedback is given in the form of an in-game text console primarily, but interface elements like labels, buttons or other text containers provide information and keep track of the user's data.

The behaviour of certain buttons may also change as the user progresses, switches activities or selects an option through a different button.

There is potential for more precise control of certain things via text input in future. 


# Game Overview 
- 14 Skills to train
- 150+ Unique enemies to fight
- 100+ NPCs to interact with, including multiple quests to complete
- 250+ Items to collect 

All of which is spread over more than 150 areas ready to be explored. 

Try playing in different ways from the start by taking advantage of the multi-account support and creating many new characters. 

Jump straight into an in-depth tutorial which will get you up to speed, or brave the unknown with the option to skip it entirely (coming soon™).


# Skills
Skills can be split into 3 main categories: Combat, Gathering, and Production. The skills within each category are generally trained in a similar fashion. Combat skills require you to kill things, with the specific skill depending on weapon and style choices. Gathering skills are trained by collecting raw resources from trees, rocks or fishing. Production skills generally process the items gained from gathering, or otherwise obtained from shops or combat, and create new potentially useful items.

| ![Skill_list](https://user-images.githubusercontent.com/96324587/147042795-df45c086-91c6-46fc-a399-a1785dd97b24.PNG) | 
|:--:| 
| *Cutting a tree while viewing the skills tab.* |

The underlying system is the same for each of the gathering skills, and the same is true for combat. Production skills tend to be more complex in design, differing from each other in interface style and certain training options.

At its core 'skilling' is simply the process of performing an action, gaining experience in a given skill for that action, and then eventually doing this enough to reach an experience milestone corresponding to a new level. Levelling up unlocks new things to do with a skill, and may also unlock access to further areas or quests.


# Items
There are lots of different kinds of items to find. Some can be equipped for use in combat while others may be used as tools to improve gathering skills. Regardless, all items have a set of actions that can be performed related to them via options in the inventory, and can be transferred to and from the bank if desired. 

| ![Inv_example](https://user-images.githubusercontent.com/96324587/147043322-c72f4b8b-505e-4b9f-a3e9-b92aeb24c08f.PNG) | 
|:--:| 
| *Looking at the inventory with logs selected.* |

Items may be uniquely obtained through skills, enemies, or even bought from shops with gold in some cases. Collecting them is important to progress, since they are vital for training production skills, completing quests and preparing for harder combat encounters.


# Areas
The game world is split into areas, each containing a unique combination of enemies, skilling locations and NPCs. They are also themed with basic background colours to add to the experience. Each area offers a selection of facilities like a bank to store items or an anvil to smith metal.

| ![Area_example](https://user-images.githubusercontent.com/96324587/147043601-585fe28f-d7e5-4b11-bc28-ea1c35f84553.PNG) | 
|:--:| 
| *Trying to smith inside a city with an anvil, bank, and cooking range.* |

Navigating between areas is simple: there are up to four directionally labeled buttons placed on appropriate sides of the window when a neighbouring area exists. In some areas there is an alternative way to travel, and an extra button will appear offering access to a non-neighbouring area. This form of transport may require certain items or payment to use.


# Combat
Each area may contain a few types of enemy to fight. Combat allows you to level combat skills and often find valuable items as a reward. The world provides a wide range of enemy difficulties to give a sense of progression, with areas generally increasing in difficulty as you move further from the starting location. It is key to note that most high end combat equipment is uniquely obtained through difficult enemy encounters.


| ![Combat_example](https://user-images.githubusercontent.com/96324587/147045032-1656c3c3-2a87-494c-b178-7399f6478ed7.PNG) | 
|:--:| 
| *Fighting a Scorpion with a melee weapon.* |

# NPCs + Quests
NPCs = Non-player characters. These exist for you to talk to, trade with and complete quests. NPC owned shops allow you to access important items easily and obtain early upgrades much faster than production skills would allow, in exchange for gold. 

| ![NPC_quest](https://user-images.githubusercontent.com/96324587/147043909-77c7b290-375d-4abc-8357-9b9d92f8b46d.PNG) | 
|:--:| 
| *Talking to an NPC about a quest.* |

Quests, while limited in scope, provide something to do as a break from training skills. For example, an NPC might ask you to kill a certain enemy or obtain a special item for them. After completing a series of these tasks, a reward is given which may include items, skill experience or access to a new area.


# Tutorial
Upon starting the game, a choice is given to register or login. If registration is chosen, then after entering a name and password the tutorial will begin. 

| [<img src="https://user-images.githubusercontent.com/96324587/147043985-11039c54-05df-4a4a-bab8-090c05da05f7.PNG" width="100%">](https://youtu.be/le1BLLInPfU) | 
|:--:| 
| *A sample play-through video of the tutorial. (Click to play)* |

The tutorial explains every major aspect of the game, walking through an example of each activity. Interface elements are introduced slowly to avoid overburdening a new user. The tutorial builds up to a point where the user is given a basic quest to complete by themselves, encompassing most of the content, which unlocks the main game upon completion.


changelog.md
# Changelog
The following is a rough list of updates in chronological order. Some things might be missed since I didn't update it too often. They start around the time the core combat system was added.

------

Swapped experience generation to after each attack to fit better with weapon and attack style switching while in combat, instead of only calculating xp on kill.\
Also reworked the combat process to account for changed attack speed and damage via weapon switching. Now checks weapon data after each attack, and can benefit from a quick to slow weapon switch for a single attack for example.

Added a bunch of new items for equipment testing.\
Added a 1hand / 2hand flag for weapons to prevent equipping shields with 2h weapons etc.\
Tested equip/inventory stuff eg. 2h weapons forces shield to be removed and prevents unequipping if full inventory.\
Added total equipment stats to equipment interface.

Completely changed inventory algorithm: now moves each item over by one slot after removing an item to fill the space.\
Added functionality to keep inventory visible while restarting combat or skills multiple times for convenience.

Reworked combat formulas.\
Added skilling success formulas, skilling tools that improve these chances and added inventory checks before skilling to find highest tier tool.\
Added Progress bar for skilling.

Added bank, using inventory objects as tabs to create 4 total bank tabs.\
Bank replaces most of the top console with the bank interface and adds quantity buttons for 1, 5, 10, 50, ALL.

Added skills interface, which displays current skill levels and updates whenever they gain experience.\
Added trade button to npc interface when selecting a shop NPC, which allows user to open the shop interface.

In bank interface: inventory item behaviour changed to deposit on left click without showing the usual options.\
Blocked bank opening while doing something since it changes the interface drastically. Inventory is now forced open during banking.\
Added bank specific add_to_inv functions and similar, allowing stacking of all items automatically.\
Added shop, similar to bank: shows an inventory style ui, but without all 28 slots if not in use. Uses same quantity selectors as bank and shows prices on the item buttons.\
Added deposit/sell all button, and last selected quantity is now saved until logout for bank and shop.

Designed magic combat using runes and requiring a staff to cast spells. User's weapon now easily determines combat style.\
Updated attack function with correct rune usage and such.\
Added Spellbook tab to the interface, showing available combat and non-combat spells with level requirements and rune costs.\
Added runes to the current test shop so they can now be purchased.

Added way to cook food, either by lighting a fire first or visiting an area containing a cooking range.\
Added way to select food in inventory as primary. Eat button will consume the selected food when pressed.

Swapped some manual item removals for the new user.inventory.remove_item() method.\
Top left part of interface now has furnace/anvil/range buttons etc. in preparation for production skills. 

Made furnace/anvil area specific rather than always available.\
Area buttons now only show up if there is a neighbouring area in the relevant direction.\
Created setup_area: sets up the interface to show only buttons that are relevant to current area.\
Added command to direction buttons to call swap_area when one exists.\
Changed direction button commands so they can now only be used when idle.

Reworked eating to work directly from inventory rather than selecting food and then pressing a separate eat button.

Added smithing, consisting of separate smelting and smithing processes with unique interfaces.\
Added hammer requirement to smithing (not the smelting portion).

Added fletching with various bows, arrows etc. Requires a knife for certain actions.\
Changed bowstrings to be stackable for ease of use.\
Added knife/hammer/bowstring etc. to basic shop so skills can be trained properly without relying on monster drops immediately.\
Now shows selected inventory item above the item's options.\
Now auto-selects next copy of same item after drop/burn to allow smooth multi-drop/burn without repeatedly selecting the item.

Added system for basic quests: simple dialogue with NPCs, checks for items in inventory on completion, tracks monster kills, requires certain skill levels to start/finish a quest, provides rewards.

Added natural HP-regen over time for QOL.\
Split up NPCs, enemies etc. into their own files to import for cleaner main file, before mass producing more of each type of object.

Finished area_setup, can now be used with direction buttons after logging in or during the tutorial. 

Added a tutorial which slowly introduces basic features, runs through an example of each skill / combat style, and then guides the user through a basic quest.

Fixed a lot of shop bugs such as the selected item not being cleared after a sale in certain situations. This led to infinite gold.\
Added daemon flag to ALL threads so there are never any errors of this type on closing. Checked each thread to make sure this doesn't break anything. 

Fully tested tutorial and the new checkpoint system. Rebalanced goblins since they were too challenging for a tutorial enemy. Possibly need to nerf final boss "Man".\
Fixed console having an inconsistent scroll position on bank/shop opening.

Added minor death mechanic: teleports user to spawnpoint on death, may change later to something more meaningful.

Added primitive save/load features via text file.\
Changed save/load methods for custom objects so that they export as a string that can be used to reconstruct them, eg. the inventory saves as Inventory([0, 1], ... ).\
Added register / login interfaces, with fully functional account creation / login leading into the tutorial / current area.


skill_obj_list.py
"""Contains all the skill objects that are placed in various areas."""

from objects import *

# Object = SkillObj(name, skill, req, xp, rate_1, rate_99, depletion_chance, resources, tool)
Tree = SkillObj("Tree", "Woodcutting", 1, 25, 65, 200, 0.125, 3, 'Axe')
Oak_tree = SkillObj("Oak tree", "Woodcutting", 15, 40, 33, 100, 0.125, 4, 'Axe')
Willow_tree = SkillObj("Willow tree", "Woodcutting", 30, 70, 15, 51, 0.125, 5, 'Axe')
Teak_tree = SkillObj("Teak tree", "Woodcutting", 35, 90, 14, 47, 0.125, 6, 'Axe')
Maple_tree = SkillObj("Maple tree", "Woodcutting", 45, 110, 8, 26, 0.125, 7, 'Axe')
Mahogany_tree = SkillObj("Mahogany tree", "Woodcutting", 50, 140, 8, 25, 0.125, 8, 'Axe')
Yew_tree = SkillObj("Yew tree", "Woodcutting", 60, 200, 0, 15, 0.125, 9, 'Axe')
Magic_tree = SkillObj("Magic tree", "Woodcutting", 75, 300, -10, 8, 0.125, 10, 'Axe')

Copper_rock = SkillObj("Copper rock", "Mining", 1, 30, 65, 200, 0.5, 77, 'Pickaxe')
Tin_rock = SkillObj("Tin rock", "Mining", 1, 30, 65, 200, 0.5, 78, 'Pickaxe')
Iron_rock = SkillObj("Iron rock", "Mining", 15, 50, 33, 100, 0.5, 89, 'Pickaxe')
Coal_rock = SkillObj("Coal rock", "Mining", 30, 60, 15, 51, 0.5, 100, 'Pickaxe')
Mithril_rock = SkillObj("Mithril rock", "Mining", 45, 100, 8, 26, 0.5, 111, 'Pickaxe')
Adamant_rock = SkillObj("Adamant rock", "Mining", 60, 150, 0, 15, 0.5, 122, 'Pickaxe')
Runite_rock = SkillObj("Runite rock", "Mining", 75, 200, -10, 8, 0.5, 133, 'Pickaxe')

Shrimp = SkillObj("Shrimp", "Fishing", 1, 15, 50, 80, 0.1, 11, 'Fishing net/rod')
Herring = SkillObj("Herring", "Fishing", 5, 30, 30, 70, 0.1, 13, 'Fishing net/rod')
Trout = SkillObj("Trout", "Fishing", 20, 45, 30, 55, 0.1, 15, 'Fishing net/rod')
Salmon = SkillObj("Salmon", "Fishing", 30, 60, 15, 50, 0.1, 17, 'Fishing net/rod')
Lobster = SkillObj("Lobster", "Fishing", 40, 90, 5, 30, 0.1, 19, 'Fishing net/rod')
Swordfish = SkillObj("Swordfish", "Fishing", 50, 120, 5, 25, 0.1, 21, 'Fishing net/rod')
Shark = SkillObj("Shark", "Fishing", 70, 180, 0, 15, 0.1, 23, 'Fishing net/rod')
Manta_ray = SkillObj("Manta ray", "Fishing", 85, 225, -5, 10, 0.1, 25, 'Fishing net/rod')


Skill_objects = {0: Tree,
                 1: Oak_tree,
                 2: Willow_tree,
                 3: Teak_tree,
                 4: Maple_tree,
                 5: Mahogany_tree,
                 6: Yew_tree,
                 7: Magic_tree,
                 8: Copper_rock,
                 9: Tin_rock,
                 10: Iron_rock,
                 11: Coal_rock,
                 12: Mithril_rock,
                 13: Adamant_rock,
                 14: Runite_rock,
                 15: Shrimp,
                 16: Herring,
                 17: Trout,
                 18: Salmon,
                 19: Lobster,
                 20: Swordfish,
                 21: Shark,
                 22: Manta_ray}


quest_list.py
"""Contains all the quests and their objectives that are used in the game."""

from objects import *

# The quests are built from a series of objectives
obj0 = Objective(0, 'item', "Oh dear, I really am in a bit of trouble. Would you be interested in a short quest?\nI "
                            "just need a few ingredients:\n1 Giant egg, 1 Bucket of milk, 1 Pot of flour",
                 "How goes it?\nI'm really pushing the deadlines here.", "I'll do it.", "No thanks.",
                 "I have them here!", "Still working on it.", 1, 1, itemsdict={276: 1, 2: 1, 38: 1})
obj1 = Objective(1, 'kill', "You're a hero!\nIf it isn't too much trouble, there's also a bit of a rat problem in the "
                            "kitchen.\nIf you kill 5 of them before people notice, the Duke will reward you for sure!",
                 "You've been helping the cook?\nI suppose there might be a reward here for you when you're done.",
                 "Alright.", "Ew! No way.", "All done sir!", "I'll be back.", 1, 0, enemiesdict={2: 5})
Cooks_assistant = Quest("Cooks Assistant", 0, None, None, "Cook", {0: 50}, {'Cooking': 500},
                        "Thank you so much for the help!\n I can teach you a small bit about cooking, "
                        "and please take some coins for your trouble.", obj0, obj1)

obj0 = Objective(0, 'kill', "Hey there! I'm interested in becoming an adventurer, like you. Could I watch you kill a "
                 "few goblins? 5 should be enough.", "Is that 5 goblins yet? I haven't been counting...",
                 "I'll do it.", "Not right now.",
                 "That's 5 of them.", "Still working on it.", 3, 3, enemiesdict={0: 5})
obj1 = Objective(1, 'item', "I think I'm ready to try some fighting, but I don't have any equipment. Could you grab a "
                 "few things for me?\nA Shortbow, 60 Bronze arrows and 5 cooked Shrimp should suffice.", "Do you have "
                 "the items? You can check your quest log if you forgot.", "I'll be right back.", "Maybe later.",
                 "I have them all here.", "Not quite yet.", 3, 3, itemsdict={151: 1, 144: 60, 12: 5})
obj2 = Objective(2, 'item', "Before I head off on adventures, could you buy some flowers and take them to my wife? She "
                 "lives on the western part of the island, and isn't too supportive of my choice.", "Hello there, is "
                 "there something you need?", "Alright.", "I'll consider it.", "I have a\ngift for you.", "No, sorry.",
                 3, 5, itemsdict={201: 1})
obj3 = Objective(3, 'kill', "I can't believe him. If you beat him in a fight, he might reconsider. Would you do that "
                 "for me?", "Has he changed his mind?", "I'll try.", "Sounds dangerous.", "Yep!", "Not just yet.",
                 5, 5, enemiesdict={4: 1})
Tut_quest = Quest("Tutorial", 1, None, None, "Man", None, {'Attack': 100, 'Defence': 100, 'Strength': 100,
                                                           'Ranged': 100, 'Magic': 100, 'Hitpoints': 100},
                  "Thank you kind stranger. Hopefully he will come home again soon. I'm sure you learned a small bit "
                  "about combat in the process, and I believe the boat to the mainland has arrived."
                  "\nGood luck on your travels.", obj0, obj1, obj2, obj3)

obj0 = Objective(0, '', "Hello there! I'm having a problem with a particular ghost in the church, and would be very "
                        "grateful if you could lend a hand.\nI need someone to get an amulet from my friend in the "
                        "swamps, Father Urney. Interested?", "What do you want then?", "I'll go find him.",
                 "Sorry, I'm busy.", "Father Eric sent me.", "Nothing.", 6, 43)
obj1 = Objective(1, 'item', "He wants the amulet doesn't he? I have it here. It allows a spiritual man to speak to "
                            "ghosts. You should get going, before I change my mind.", "Do you have it? This thing is "
                            "driving me insane!", "I'm on it.", "Maybe.", "Got it right here.", "Sorry, not yet.",
                 43, 6, itemsdict={281: 1}, quest_items=[281])
obj2 = Objective(2, 'item', "Okay, the ghost seems to be angry that his body's skull is missing. He saw a wizard take "
                            "it, presumably for some illegal necromancy. Try searching the wizard's tower to the west, "
                            "there may be a re-animated skeleton making use of the head. Get it back.",
                 "Got the head? I'll put it back in the coffin for you, and this ghost can rest at last.",
                 "I'll take a look.", "Sounds a bit too scary.", "Here, catch!", "I'm still looking.", 6, 6,
                 itemsdict={282: 1})
Restless_ghost = Quest("Restless ghost", 2, None, None, "Father Eric", {0: 250}, {'Hitpoints': 250},
                       "Thanks! Hopefully that's the end of the noise. Take this gold for your troubles.", obj0, obj1,
                       obj2)
obj0 = Objective(0, '', "Welcome, adventurer. You look like someone who enjoys a quest now and again. Would you do me "
                        "a favour and check on our man in the temple, to the east? He is tasked with guarding us from "
                        "the evils of the far east, but I haven't heard from Drazel for a week now.", "Pssst, I'm over "
                        "here, in the jail cell! Did the king send you?", "I do love a good quest!", "No thanks.",
                        "Yes, I'm here to help.", "No, just watching.", 23, 26)
obj1 = Objective(1, '', "These evil monks have broken into the temple and locked me up. I'm sure one of them has the "
                        "key, see if you can get it for me.", "Did you find it?", "One key coming right up...",
                        "Not my problem.", "Yep!", "Not yet.", 26, 26, itemsdict={283: 1})
obj2 = Objective(2, '', "Ahh, that's better. Unfortunately, they destroyed the holy statue protecting us from the east."
                        " I'll need some un-focused rune stones to repair it, but that kind of stuff isn't easy to "
                        "find. Could you talk to Treyborn at the wizard's tower? He's an old friend.",
                 "What brings you to the tower, stranger? I'm having a terrible day.", "I'll talk to him.",
                 "Sounds like work.", "I need some \nun-focused rune stones.", "Just browsing.", 26, 44)
obj3 = Objective(3, '', "Rune stones? I shouldn't be sharing those with a non-wizard. However, I owe Drazel a favour, "
                        "so maybe I can help you out. Some imps have stolen my magic beads, and ran off to the north. "
                        "If you can find all 4, I'll give you what you want.", "Have you found my beads?",
                 "Sounds like a deal.", "Screw your beads, old man.", "Got them right here.", "Nope.", 44, 44,
                 itemsdict={277: 1, 278: 1, 279: 1, 280: 1})
obj4 = Objective(4, '', "Alright, a deal's a deal. Take these stones, but don't give them to anyone except Drazel. "
                        "They are very powerful.", "Did he give you the stones? I'm not sure how long we have until the"
                                                   " evil starts creeping in.",
                 "Thanks!", "About time.", "He did, but it wasn't easy.", "No, still trying.", 44, 26,
                 itemsdict={284: 1}, quest_items=[284])
obj5 = Objective(5, '', "Oh good! Thank you, traveller. I'll have the statue rebuilt shortly. You can run along and "
                        "tell the king the news.", "Do you have news from the east?", "Bye!", "Later.",
                 "Drazel was in trouble,\nbut everything is alright now.", "Nothing new.", 26, 23)
Priest_in_peril = Quest('Priest in peril', 3, None, [2], "King Roald", {0: 2500},
                        {'Attack': 1000, 'Strength': 1000, 'Defence': 1000, 'Hitpoints': 1000, 'Ranged': 1000,
                         'Magic': 1000},
                        "Good work. Take some gold from my treasury. I'll also tell Drazel to allow you passage "
                        "through the temple, if you ever want to help our fight in the east.",
                        obj0, obj1, obj2, obj3, obj4, obj5)
obj0 = Objective(0, '', "Will you help me? I'm in love with a girl, Juliet, but our families forbid us from seeing "
                        "each other. I'm desperate to see her, but I'm out of ideas. Could you talk to her for me? "
                        "Maybe she has a plan.", "What are you looking at?", "Anything for love.", "Not interested.",
                 "Romeo sent me.", "You, stupid.", 22, 41)
obj1 = Objective(1, '', "Oh, Romeo. I was working on something that could get us together once and for all, "
                        "but I can't do it alone. The details are in this letter. If you show it to Romeo, he can "
                        "work out the rest. Are you up for it?", "So, did she have any ideas?",
                 "I guess I'm already helping.", "I'll consider it.", "She gave me this letter.", "I'll ask again.",
                 41, 22, itemsdict={285: 1}, quest_items=[285])
obj2 = Objective(2, '', "Oh, this is brilliant! Looks like she wants to fake her death, then I can meet her in the "
                        "crypts and we can escape together. We need some sort of way to make it believable though. "
                        "Father Lawrence is a trusted friend, could you try talking to him for me?",
                 "Welcome to the church. Do you need something in particular?", "I'll go find him.",
                 "This is getting complicated.", "Romeo needs help.", "Just praying.", 22, 111)
obj3 = Objective(3, '', "Faking her death? I heard of a special potion that could work for that purpose. The apothecary"
                        " would know more, though. Give him a visit.",
                 "I make all sorts of potions, but nobody seems to use them. Do you want one?",
                 "I'll see what he knows.", "I'm done helping.", "Yes, a very special one.", "Keep your potions.",
                 111, 42)
obj4 = Objective(4, '', "Ah, a customer! I do know how to make such a potion, but I don't keep ingredients on hand. I "
                        "need special berries, Cadava, that usually grow outside the city to the southeast. Bears tend "
                        "to steal them though, so you might need to kill a few.",
                 "Do you have them? I hope it wasn't much trouble.", "I'll be back.", "I'm not your errand boy.",
                 "Got them here.", "Working on it.", 42, 42, itemsdict={286: 1})
obj5 = Objective(5, '', "Just a mix here, a quick shake there, and... Done! One sip of this baby and she'll be out like"
                        " a light. Any local doctor will think she's dead, at least for a while. Take this to her.",
                 "Did you think of something? I can't wait much longer.", "I'll get right on it.", "About time.",
                 "Yep! This potion should work.", "No clue.", 42, 41, quest_items=[287], itemsdict={287: 1})
obj6 = Objective(6, '', "A potion? Are you sure it's safe? Whatever, I'd rather die than never see Romeo again. Tell "
                        "him to meet me in the crypts, tonight. Hopefully I'll still be breathing.",
                 "Is everything set up? I can't wait longer.", "On my way!", "Good luck.",
                 "Yes, in the crypts tonight.", "Not yet.", 41, 22)
Romeo_and_juliet = Quest('Romeo and Juliet', 4, None, None, 'Romeo', {0: 500}, None,
                         "Oh good! Here, take some gold for your troubles. I won't make you wait around for tonight. "
                         "From here on out it's just me and Juliet. Thanks so much!",
                         obj0, obj1, obj2, obj3, obj4, obj5, obj6)

Quests = {0: Cooks_assistant,
          1: Tut_quest,
          2: Restless_ghost,
          3: Priest_in_peril,
          4: Romeo_and_juliet}
